polkadot/runtime/parachains/src/broadcaster/mod.rs

```diff
@@ -0,0 +1,397 @@
+// Copyright (C) Parity Technologies (UK) Ltd.
+// This file is part of Polkadot.
+
+// Polkadot is free software: you can redistribute it and/or modify
+// it under the terms of the GNU General Public License as published by
+// the Free Software Foundation, either version 3 of the License, or
+// (at your option) any later version.
+
+// Polkadot is distributed in the hope that it will be useful,
+// but WITHOUT ANY WARRANTY; without even the implied warranty of
+// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+// GNU General Public License for more details.
+
+// You should have received a copy of the GNU General Public License
+// along with Polkadot.  If not, see <http://www.gnu.org/licenses/>.
+
+//! A pallet for managing parachain data publishing and subscription.
+//!
+//! This pallet provides a publish-subscribe mechanism for parachains to share data
+//! efficiently through the relay chain storage using child tries per publisher.
+//!
+//! ## Storage Lifecycle
+//!
+//! Note: This pallet does not currently implement publisher removal or cleanup mechanisms.
+//! Once a parachain publishes data, it remains in storage. Publishers can update their data
+//! by publishing again, but there is no explicit removal path.
+
+use alloc::{collections::BTreeMap, vec::Vec};
+use frame_support::{
+	pallet_prelude::*,
+	storage::child::ChildInfo,
+	traits::{defensive_prelude::*, Get, ConstU32},
+};
+use frame_system::pallet_prelude::BlockNumberFor;
+use polkadot_primitives::Id as ParaId;
+
+pub use pallet::*;
+
+mod traits;
+pub use traits::PublishSubscribe;
+
+#[cfg(test)]
+mod tests;
+
+#[frame_support::pallet]
+pub mod pallet {
```

Comment:
There is no logic that cleans up the data after a parachain got off-boarded.

We probably also request that parachains first need to register themselves for publishing data (if they are not a system chain) and require a deposit for the ~2KiB of data.